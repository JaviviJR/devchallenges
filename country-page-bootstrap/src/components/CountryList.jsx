import PropTypes from 'prop-types';
import styles from './CountryList.module.css';
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';

const ITEMS_PER_PAGE = 10;

function CountryList({ countries }) {
    const navigate = useNavigate();
    const [currentPage, setCurrentPage] = useState(1);

    const totalPages = Math.ceil(countries.length / ITEMS_PER_PAGE);

    const currentCountries = countries.slice(
        (currentPage - 1) * ITEMS_PER_PAGE,
        currentPage * ITEMS_PER_PAGE
    );

    const handleNextPage = () => {
        setCurrentPage((page) => Math.min(page + 1, totalPages));
    };

    const handlePreviousPage = () => {
        setCurrentPage((page) => Math.max(page - 1, 1));
    };

    const handleClickedCountry = (id) => {
        navigate(`/country/${id}`);
    };

    return (
        <div className={`${styles.countryList}`}>
            <table className={`table align-middle`}>
                <thead>
                    <tr>
                        <th scope="col">Flag</th>
                        <th scope="col">Name</th>
                        <th scope="col">Population</th>
                        <th scope="col">Area (km2)</th>
                        <th scope="col" className="d-none d-lg-table-cell">Region</th>
                    </tr>
                </thead>
                <tbody className="table-group-divider" style={{borderTop: "1px solid #272a2f"}}>
                    { currentCountries.map((country, index) => (
                        <tr key={ index } onClick={ () => handleClickedCountry(country.ccn3) }>
                            <td><img className={styles.flagImage} src={country.flag} /></td>
                            <td>{ country.name }</td>
                            <td>{ country.population.toLocaleString() }</td>
                            <td>{ country.area.toLocaleString() }</td>
                            <td className="d-none d-lg-table-cell">{ country.region }</td>
                        </tr>
                    ))}
                </tbody>
            </table>
            <div className={`${styles.pagination}`}>
                <button className={`${styles.paginationButton}`} onClick={handlePreviousPage} disabled={currentPage === 1}>Previous</button>
                <span className={`${styles.paginationInfo}`}>Page {currentPage} of {totalPages}</span>
                <button className={`${styles.paginationButton}`} onClick={handleNextPage} disabled={currentPage === totalPages}>Next</button>
            </div>
        </div>
    );
}

export default CountryList;

CountryList.propTypes = {
    countries: PropTypes.array.isRequired
}