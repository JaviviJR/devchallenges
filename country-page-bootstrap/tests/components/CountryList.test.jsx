import { render, screen, within } from "@testing-library/react"
import userEvent from '@testing-library/user-event'
import CountryList from "../../src/components/CountryList";

const countries = Array.from({ length: 20 }, (_, i) => (
    {
        ccn3: `ABC${i}`,
        name: `Country ${i + 1}`,
        population: `${(i + 1) * 10000}`,
        area: `${(i + 1)*10}`,
        region: `regionTest${i}`,
        subregion: `subregionTest${i}`,
        flag: `https://example.com/flags/test${i}.svg`,
        independent: true,
        unMember: true,
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => jest.fn(),
}));

describe('testing CountryList component', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should match with snapshot', () => {
        const { container } = render(
            <CountryList countries={ countries }/>
        );

        expect(container).toMatchSnapshot();
    })

    test('should render table and pagination', () => {
        render(
            <CountryList countries={ countries }/>
        );
        
        const table = screen.getByRole('table');
        const rows = screen.getAllByRole('row');
        const nextButton = screen.getByRole('button', {name: 'Next'});
        const previousButton = screen.getByRole('button', {name: 'Next'});

        expect(table).toBeInTheDocument();
        expect(rows.length).toEqual(11); // +1 from row header
        expect(nextButton).toBeInTheDocument();
        expect(previousButton).toBeInTheDocument();
    });

    test('should paginate country list', () => {
        render(
            <CountryList countries={ countries }/>
        );
        const rows = screen.getAllByRole('row');
        expect(rows.length).toEqual(11); // +1 from row header
    })

    test('should displays different countries on next page', async () => {
        const user = userEvent.setup();
      
        render(<CountryList countries={countries} />);
      
        const firstPageRows = screen.getAllByRole('row');        
        let cells = within(firstPageRows[1]).getAllByRole('cell');        
        const firstCountryName = cells[1].innerHTML;

        const nextButton = screen.getByRole('button', {name: 'Next'});
        await user.click(nextButton);
        
        const secondPageRows = screen.getAllByRole('row');        
        cells = within(secondPageRows[1]).getAllByRole('cell');        
        const secondCountryName = cells[1].innerHTML;
      
        expect(firstCountryName).not.toEqual(secondCountryName);
      });
})