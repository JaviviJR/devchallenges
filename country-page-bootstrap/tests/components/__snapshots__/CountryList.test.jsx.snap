// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing CountryList component should match with snapshot 1`] = `
<div>
  <div
    class="countryList"
  >
    <table
      class="table align-middle"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            Flag
          </th>
          <th
            scope="col"
          >
            Name
          </th>
          <th
            scope="col"
          >
            Population
          </th>
          <th
            scope="col"
          >
            Area (km2)
          </th>
          <th
            class="d-none d-lg-table-cell"
            scope="col"
          >
            Region
          </th>
        </tr>
      </thead>
      <tbody
        class="table-group-divider"
        style="border-top: 1px solid #272a2f;"
      >
        <tr>
          <td>
            <img
              class="flagImage"
              src="https://example.com/flags/test0.svg"
            />
          </td>
          <td>
            Country 1
          </td>
          <td>
            10000
          </td>
          <td>
            10
          </td>
          <td
            class="d-none d-lg-table-cell"
          >
            regionTest0
          </td>
        </tr>
        <tr>
          <td>
            <img
              class="flagImage"
              src="https://example.com/flags/test1.svg"
            />
          </td>
          <td>
            Country 2
          </td>
          <td>
            20000
          </td>
          <td>
            20
          </td>
          <td
            class="d-none d-lg-table-cell"
          >
            regionTest1
          </td>
        </tr>
        <tr>
          <td>
            <img
              class="flagImage"
              src="https://example.com/flags/test2.svg"
            />
          </td>
          <td>
            Country 3
          </td>
          <td>
            30000
          </td>
          <td>
            30
          </td>
          <td
            class="d-none d-lg-table-cell"
          >
            regionTest2
          </td>
        </tr>
        <tr>
          <td>
            <img
              class="flagImage"
              src="https://example.com/flags/test3.svg"
            />
          </td>
          <td>
            Country 4
          </td>
          <td>
            40000
          </td>
          <td>
            40
          </td>
          <td
            class="d-none d-lg-table-cell"
          >
            regionTest3
          </td>
        </tr>
        <tr>
          <td>
            <img
              class="flagImage"
              src="https://example.com/flags/test4.svg"
            />
          </td>
          <td>
            Country 5
          </td>
          <td>
            50000
          </td>
          <td>
            50
          </td>
          <td
            class="d-none d-lg-table-cell"
          >
            regionTest4
          </td>
        </tr>
        <tr>
          <td>
            <img
              class="flagImage"
              src="https://example.com/flags/test5.svg"
            />
          </td>
          <td>
            Country 6
          </td>
          <td>
            60000
          </td>
          <td>
            60
          </td>
          <td
            class="d-none d-lg-table-cell"
          >
            regionTest5
          </td>
        </tr>
        <tr>
          <td>
            <img
              class="flagImage"
              src="https://example.com/flags/test6.svg"
            />
          </td>
          <td>
            Country 7
          </td>
          <td>
            70000
          </td>
          <td>
            70
          </td>
          <td
            class="d-none d-lg-table-cell"
          >
            regionTest6
          </td>
        </tr>
        <tr>
          <td>
            <img
              class="flagImage"
              src="https://example.com/flags/test7.svg"
            />
          </td>
          <td>
            Country 8
          </td>
          <td>
            80000
          </td>
          <td>
            80
          </td>
          <td
            class="d-none d-lg-table-cell"
          >
            regionTest7
          </td>
        </tr>
        <tr>
          <td>
            <img
              class="flagImage"
              src="https://example.com/flags/test8.svg"
            />
          </td>
          <td>
            Country 9
          </td>
          <td>
            90000
          </td>
          <td>
            90
          </td>
          <td
            class="d-none d-lg-table-cell"
          >
            regionTest8
          </td>
        </tr>
        <tr>
          <td>
            <img
              class="flagImage"
              src="https://example.com/flags/test9.svg"
            />
          </td>
          <td>
            Country 10
          </td>
          <td>
            100000
          </td>
          <td>
            100
          </td>
          <td
            class="d-none d-lg-table-cell"
          >
            regionTest9
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="pagination"
    >
      <button
        class="paginationButton"
        disabled=""
      >
        Previous
      </button>
      <span
        class="paginationInfo"
      >
        Page 
        1
         of 
        2
      </span>
      <button
        class="paginationButton"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
